### ISSUE TITLE
EINTEGRITY error doesn't say for which module

### ISSUE THREAD

**Issue by vjpr** [pnpm/pnpm#872]
```
  27393 debug pnpm:progress:
    status: "downloaded_manifest"
    pkgId: "registry.npmjs.org/draft-js/0.10.1"
    pkgVersion: "0.10.1"
  27394 debug pnpm:progress:
    status: "dependencies_installed"
    pkgId: "registry.npmjs.org/draft-js/0.10.1"
  27395 debug pnpm:progress:
    status: "downloaded_manifest"
    pkgId: "registry.npmjs.org/fbjs/0.8.14"
    pkgVersion: "0.8.14"
  27396 debug pnpm:progress:
    status: "dependencies_installed"
    pkgId: "registry.npmjs.org/fbjs/0.8.14"
  27397 debug pnpm:progress:
    status: "downloaded_manifest"
    pkgId: "registry.npmjs.org/object-assign/4.1.1"
    pkgVersion: "4.1.1"
  27398 debug pnpm:progress:
    status: "dependencies_installed"
    pkgId: "registry.npmjs.org/object-assign/4.1.1"
  27399 debug pnpm:progress:
    status: "dependencies_installed"
    pkgId: "registry.npmjs.org/rc-editor-core/0.7.8"
  27400 debug pnpm:fetch: "finish sha1-E7TTyxK++hVIKib+Gy665kAHHks= https://registry.npmjs.org/immutable/-/immutable-3.7.6.tgz"
  27401 debug pnpm:progress:
    status: "fetched"
    pkgId: "registry.npmjs.org/immutable/3.7.6"
  27402 debug pnpm:progress:
    status: "downloaded_manifest"
    pkgId: "registry.npmjs.org/immutable/3.7.6"
    pkgVersion: "3.7.6"
  27403 debug pnpm:progress:
    status: "dependencies_installed"
    pkgId: "registry.npmjs.org/immutable/3.7.6"
  27404 debug pnpm:progress:
    status: "dependencies_installed"
    pkgId: "registry.npmjs.org/draft-js/0.10.1"
  27405 debug pnpm:progress:
    status: "dependencies_installed"
    pkgId: "registry.npmjs.org/rc-editor-mention/0.6.12"
  27406 debug pnpm:progress:
    status: "dependencies_installed"
    pkgId: "registry.npmjs.org/antd/2.12.5"
  27407 error pnpm:
    message:
      code: "EINTEGRITY"
      found: "sha1-dbSmLq6UB++20xCT5eJrkj11178="
      expected:
        - "sha1-mnHEh0chjrylHlGmbaaCA4zct78="
      algorithm: "sha1"
      sri: "sha1-mnHEh0chjrylHlGmbaaCA4zct78="
    err:
      name: "Error"
      message: "sha1-mnHEh0chjrylHlGmbaaCA4zct78= integrity checksum failed when using sha1: wanted sha1-mnHEh0chjrylHlGmbaaCA4zct78= but got sha1-dbSmLq6UB++20xCT5eJrkj11178=. (4292233 bytes)"
      code: "EINTEGRITY"
      stack: "Error: sha1-mnHEh0chjrylHlGmbaaCA4zct78= integrity checksum failed when using sha1: wanted sha1-mnHEh0chjrylHlGmbaaCA4zct78= but got sha1-dbSmLq6UB++20xCT5eJrkj11178=. (4292233 bytes)\n    at Transform.on (/usr/local/lib/node_modules/pnpm/lib/node_modules/ssri/index.js:275:19)\n    at emitNone (events.js:110:20)\n    at Transform.emit (events.js:207:7)\n    at endReadableNT (_stream_readable.js:1047:12)\n    at _combinedTickCallback (internal/process/next_tick.js:102:11)\n    at process._tickCallback (internal/process/next_tick.js:161:9)"
```

For my particular case it works on macOS but not on Ubuntu.

### Additional information:

 - `node -v` prints: 8.1.2
 - Windows, OS X, or Linux?: Ubuntu 16

BTW: Is there an easier way to check which module it was?

**Comment by vjpr** [comment]
So I just had to look in the shrinkwrap.yaml.

For me the module was:

```
  /material-design-icons/3.0.1:
    resolution:
      integrity: sha1-mnHEh0chjrylHlGmbaaCA4zct78=
```

Could this integrity fail between OS's?

**Comment by vjpr** [comment]
Same issue reported with npm here: https://github.com/npm/npm/issues/16861#issuecomment-321486637

**Comment by zkochan** [comment]
The error is thrown from [here](https://github.com/pnpm/package-store/blob/master/src/network/got.ts#L125).

**Comment by vjpr** [comment]
I don't think it is about this package in particular. I am thinking its about a partial tar download occurring in `npm-package-client`.

**Comment by jbergstroem** [comment]
> @vjpr said:
> I am thinking its about a partial tar download occurring in `npm-package-client`.

I've suspect this is the root cause for similar issues I've experienced in poor wifi environments.

**Comment by zkochan** [comment]
@jbergstroem do you think we need to redownload the package if this error occurs?

**Comment by vjpr** [comment]
See more detail here: https://github.com/npm/npm/issues/16861#issuecomment-322603643

I think its an issue with `npm-registry-client`.

Re-download might be a good workaround for now though. Should definitely have message letting user know what is going on though and have a retry limit.

**Comment by vjpr** [comment]
I was just downloading a module on a patchy connection and got this error:

```
$ pnpm i -S react-vis
Creating dependency tree
Adding 32 packages to node_modules
ERROR sha1-L8eZlME0N9+o/a8iztjNFGj4GiQ= integrity checksum failed when using sha1: wanted sha1-L8eZlME0N9+o/a8iztjNFGj4Gi
Q= but got sha1-zx2Bd3m2OUeRHPqbJSs5zzDlb+M=. (2195456 bytes)
Resolving: total 77, reused 54, downloaded 22
```

Running `pnpm i react-vis` again showed a successful install, but I cannot see it in `node_modules`. This seems like a bug.

**Comment by vjpr** [comment]
I just hit this again...

```
$ ✗ lsl ~/.pnpm-store/2/registry.npmjs.org/warning/3.0.0
total 24
drwxr-xr-x  6 Vaughan  staff   204B 21 Aug 11:35 .
drwxr-xr-x  4 Vaughan  staff   136B 21 Aug 11:35 ..
-rw-r--r--  1 Vaughan  staff   1.1K 21 Aug 11:35 integrity.json
drwxr-xr-x  3 Vaughan  staff   102B 21 Aug 11:35 node_modules
lrwxr-xr-x  1 Vaughan  staff    20B 21 Aug 11:35 package -> node_modules/warning
-rw-r--r--  1 Vaughan  staff   2.9K 21 Aug 11:35 packed.tgz
$ ✗ lsl ~/.pnpm-store/2/registry.npmjs.org/material-design-icons/
total 0
drwxr-xr-x     4 Vaughan  staff   136B 21 Aug 11:35 .
drwxr-xr-x  1834 Vaughan  staff    61K 21 Aug 11:36 ..
drwxr-xr-x     3 Vaughan  staff   102B 21 Aug 11:35 3.0.1
drwxr-xr-x    19 Vaughan  staff   646B 21 Aug 11:37 3.0.1_stage
$ ✗ lsl ~/.pnpm-store/2/registry.npmjs.org/material-design-icons/3.0.1
total 26224
drwxr-xr-x  3 Vaughan  staff   102B 21 Aug 11:35 .
drwxr-xr-x  4 Vaughan  staff   136B 21 Aug 11:35 ..
-rw-r--r--  1 Vaughan  staff    13M 21 Aug 11:37 packed.tgz.3627129203
$ ✗ lsl ~/.pnpm-store/2/registry.npmjs.org/material-design-icons/3.0.1_stage
total 72
drwxr-xr-x  19 Vaughan  staff   646B 21 Aug 11:37 .
drwxr-xr-x   4 Vaughan  staff   136B 21 Aug 11:35 ..
-rwxr-xr-x   1 Vaughan  staff    13B  2 Sep  2016 .npmignore
-rw-r--r--   1 Vaughan  staff    11K  2 Sep  2016 LICENSE
-rw-r--r--   1 Vaughan  staff   1.9K  2 Sep  2016 README.md
drwxr-xr-x  12 Vaughan  staff   408B 21 Aug 11:35 alert
drwxr-xr-x  12 Vaughan  staff   408B 21 Aug 11:36 av
-rw-r--r--   1 Vaughan  staff   289B  2 Sep  2016 bower.json
drwxr-xr-x  12 Vaughan  staff   408B 21 Aug 11:36 communication
drwxr-xr-x  17 Vaughan  staff   578B 21 Aug 11:36 content
drwxr-xr-x  12 Vaughan  staff   408B 21 Aug 11:36 device
drwxr-xr-x  12 Vaughan  staff   408B 21 Aug 11:37 editor
drwxr-xr-x  12 Vaughan  staff   408B 21 Aug 11:37 file
-rwxr-xr-x   1 Vaughan  staff   3.8K  2 Sep  2016 gulpfile.babel.js
drwxr-xr-x  12 Vaughan  staff   408B 21 Aug 11:37 hardware
drwxr-xr-x  11 Vaughan  staff   374B 21 Aug 11:37 iconfont
drwxr-xr-x   6 Vaughan  staff   204B 21 Aug 11:37 image
-rw-r--r--   1 Vaughan  staff    48B  2 Sep  2016 index.js
-rw-r--r--   1 Vaughan  staff   872B  2 Sep  2016 package.json
```

**Comment by vjpr** [comment]
So I think what is happening is that the npm registry is occasionally prematurely closing its socket for large file downloads, which is not being reported as an error by `request` http client, although I'm not even sure if it can be.

We need to add re-download logic if the integrity fails.

**Comment by vjpr** [comment]
Ok I feel like I have got to the bottom of this issue now. Details here: https://github.com/npm/registry/issues/202

For pnpm, we need to check the `Content-Length` of the response, and then retry the download.

**Comment by zkochan** [comment]
There is an old issue about Content-Length: https://github.com/pnpm/pnpm/issues/423. If we solve this, we can close that one

**Comment by vjpr** [comment]
Yep. pnpm/package-store is what needs changing I think.

**Comment by jbergstroem** [comment]
> @zkochan said:
> @jbergstroem do you think we need to redownload the package if this error occurs?

Apologies, didn't see this until now. I'd either bail early or retry. Seeing how `pnpm` is taking the Just Works approach (successfully so), I think retrying with a configurable amount of attempts makes sense.

**Comment by zkochan** [comment]
This is the solution npm will apply but they won't add it to `npm-registry-client` (they are deprecating it slowly), so we'll have to add the retry logic to our [package-store](https://github.com/pnpm/package-store) package.

The hardest part for me is to write an integration test for it

**Comment by zkochan** [comment]
Published with [pnpm@1.13.1](https://github.com/pnpm/pnpm/releases/tag/v1.13.1)
